New features

nanobowers opened the pull request:
A few new features added:

1. allow hidden options - ones that dont print with `educate`
2. "global" settings to enable behavior similar to Perl's Getopt::Long
3. prevent short-flags from being generated for each option by default
this is more convenient than setting `:short => :none` for each option
IMX, users often forget to use two dashes for long opts and then inadvertently trigger short-opts, e.g., instead of triggering option ` --dog`, they effectively trigger `-d -o -g`, or `-d og`
4. enable loose matching of options for the shortest unambiguous match.  Given options `apple`, `apricot` and `antelope`, you could trigger them with `--app, --apr, and --an`, respectively.  

Tests and some amount of documentation also included.

nanobowers commented:
@kbrock any feedback on this PR?

nanobowers commented:
Submitted this PR about 4 months ago.
Can this be approved/denied or otherwise given some feedback, please?

kbrock commented on lib/trollop.rb:
```diff
@@ -220,7 +232,7 @@ def parse(cmdline = ARGV)
     end
 
-    resolve_default_short_options!
+    resolve_default_short_options! unless @settings[:no_default_short_opts]
```
:+1:

kbrock replied:
not thrilled with this name. Are there other option names that seem similar to this one?

kbrock commented on lib/trollop.rb:
```diff
@@ -761,9 +788,12 @@ def single_arg?
   def multi ; opts[:multi] ; end
   alias multi? multi
 
+  def hidden ; opts[:hidden] ; end
```
not thrilled about adding more flags here

kbrock commented on lib/trollop.rb:
```diff
@@ -61,7 +61,11 @@ class Parser
   attr_accessor :ignore_invalid_options
 
+  ## Settings for the entire option parser that apply to all given options.
+  attr_reader :settings
```
Leaning away from this feature.
Although, I guess we do have some global behavior embedded in this gem

nanobowers commented:
Hi @kbrock, I've addressed much of your feedback with some recent commits.

- Reverted versions back to 2.1.2
- Removed attr_reader :settings .  There is no need for a public method here to retrieve the settings on the instance of Trollop::options , it was just used for testing purposes.
- Settings is documented on lines 856-867 of lib/trollop.rb
- However, there is still a need to pass some sort of settings that can be applied to the entirety of the Trollop::options call (insofar as implementing the inexact_match and no_default_short_opts flags are concerned).
- A :hidden flag on an option is useful, in my experience.  Developers of command line utilities used in small groups like to roll out undocumented options first - either for their own use, for power-users, or just for limited audience testing.
- RE: naming of no_default_short_opts flag, maybe it's poorly named, but it made sense to me.  It is used to suppress generation of short options unless they are manually specified.  Maybe "prevent_default_short" ?
- I have modified the inexact-match (lazy match) to handle the case you mentioned, which is to say that if you have options :book and :bookstand, then  --book will exact match :book (unambiguously).  Likewise --books will only match bookstand, as it matches the first portion of the given string against the existing long options.  I enhanced my collision tests a bit.

I understand the hesitation on these features. If some are unpalatable, let me know and i'll remove them.

chrisarcand commented:
> prevent short-flags from being generated for each option by default

@kbrock Note this is a backwards incompatible change to the very interface of the gem and would require a major version bump.

nanobowers commented:
@chrisarcand i'm not sure how this is backwards incompatible.  All of the previous tests work just fine without this new option enabled.  I think there may be some confusion due to how i named it: it probably should've read something like "An option (defaulted off) that will prevent short-flags from automatically generated".  perhaps that's what @kbrock meant by "not thrilled with this name"  - maybe `no_auto_shorts`? Or am i missing something about how this will break existing users?

nanobowers commented:
I think the PR should be completely backward compatible for the case you suggest, this was a case of me not clearly expressing my intent.

The aforementioned:
> prevent short-flags from being generated for each option by default
> enable loose matching of options 

Should have read something like
Add two flags: 
1. disable automatic short flag generation
2. enable loose matching of options

for the sake of backward compatibility with hundreds/thousands of existing projects, neither of these are turned on by default in this PR.  A project developer would have to intentionally throw a switch in order to alter behavior of said project.  My intent is to extend the functionality of trollop, not change what it does when normally configured.

chrisarcand commented:
Ahhh I see, my confusion was a due to a combination of just reading your description and not closely looking at the code itself, and that a comment not made by you seemed to indicate this behavior from the linked issue.

I see you have this behavior behind the `:no_default_short_opts` option. 

Forget my comments, you're right! This is good.

nanobowers commented:
latest push:
* changed confusing setting name to :disable_auto_short_opts

* added another new feature setting.  :suggestions
this uses the DidYouMean module (only when available, but it's standard in 2.3 and 2.4) to provide Git-like responses when a user gives an option name that doesn't exist.  It looks for close matches and gives suggestions.  This only affects the output message on the raised CommandlineError.  The suggestion tests are only enabled when DidYouMean is found, so the developement dependencies do not require it.

kbrock commented on lib/trollop.rb:
```diff
@@ -76,8 +77,18 @@ def initialize(*a, &b)
     # instance_eval(&b) if b # can't take arguments
-    cloaker(&b).bind(self).call(*a) if b
+    #cloaker(&b).bind(self).call(*a) if b
+    self.instance_exec(*a,&b) if block_given?
```
what is cloaker?
Is this a __why fix for a ruby 1.8.6 limitation?

kbrock replied:
We've been using methods as a way to introduce "settings"
Do we really want to change this?

stylistically, it is nice.

nanobowers replied:
The cloaker is the _why fix.  Unfortunately it doesn't seem to work properly with more complicated arg passing.  It has been months since i've last looked at that.

Using params for setting "global" settings on `Trollop.options` seemed to me a better than adding them to as methods, as the methods are sequenced in an order.  If you have a method named `disable_auto_short_options`, then you'd need to set that before any `opt` method call.  this would either require running the "settings" methods first to get a sane result.

kbrock commented on lib/trollop.rb:
```diff
@@ -203,6 +214,53 @@ def educate_on_error
+  ## If we hit a complete match, then use that, otherwise see how many long-options partially match.
+  ## If only one partially matches, then we can safely use that.
+  ## Otherwise, we raise an error that the partially given option was ambiguous.
+  def perform_inexact_match (arg, partial_match)  # :nodoc:
```
as long as this is not enabled by default, looks good

kbrock commented on lib/trollop.rb:
```diff
@@ -220,7 +278,7 @@ def parse(cmdline = ARGV)
     end
 
-    resolve_default_short_options!
+    resolve_default_short_options! unless @settings[:disable_auto_short_opts]
```
+1

kbrock commented on lib/trollop.rb:
```diff
@@ -747,6 +811,9 @@ def initialize(name, desc="", opts={}, &b)
     opts[:multi] ||= false
 
+    ## fill in :hidden
+    opts[:hidden] ||= false
```
wish we didn't need so much code for an option that overrides this gem trying to be clever (by picking the short names)

nanobowers commented:
I merged on my end with master  / latest and pushed again.  The 'globals' are not really globals (e.g. class methods), but are settings that affect all options created on the instance of the parser, e.g:
```ruby
## Settings:
##   Trollop::options and Trollop::Parser.new accept +settings+ to control how
##   options are interpreted.  These settings are given as hash arguments, e.g:
##
##   opts = Trollop::options(ARGV, :inexact_match => true, :disable_auto_short_opts => true ) do
##     opt :foobar, 'messed up'
##     opt :forget, 'forget it'
##   end
##
##  +settings+ include:
##  * :inexact_match  : Allow minimum unambigous number of characters to match a long option
##  * :disable_auto_short_opts  : Prevent automatic creation of short options
##  * :suggestions  : Enables suggestions when unknown arguments are given and DidYouMean is installed.  DidYouMean comes standard with Ruby 2.3+
##  Because Trollop::options uses a default argument for +args+, you must pass that argument when using the settings feature.
```
Maybe it's better to refer to them as 'locals' rather than 'globals'.

nanobowers commented:
Any thought on this?  These enhancements (including the suggestions / DidYouMean) are pretty useful from an end-user standpoint.
